### ISSUE TITLE
Taptap exiting right after invoking

### ISSUE THREAD

**Issue by Scorcerer** [willglynn/taptap#2]
Hi!
Thanks for the project.
I pulled your repo and compiled it as a standard rust project, but unfortunately the binary exits basically immediately after I try to connect it to my rs-485 gateway:
```
scor@Cloud ~/Projects/taptap $ cargo build
    Updating crates.io index
  Downloaded anstyle v1.0.8
  Downloaded adler v1.0.2
  Downloaded autocfg v1.3.0
  Downloaded anstyle-parse v0.2.5
  Downloaded is_terminal_polyfill v1.70.1
  Downloaded libudev-sys v0.1.4
  Downloaded cfg-if v1.0.0
  Downloaded colorchoice v1.0.2
  Downloaded anstyle-query v1.1.1
  Downloaded bitflags v1.3.2
  Downloaded anstream v0.6.15
  Downloaded env_filter v0.1.2
  Downloaded thiserror v1.0.62
  Downloaded utf8parse v0.2.2
  Downloaded zerocopy-derive v0.8.0-alpha.16
  Downloaded serialport v4.4.0
  Downloaded miniz_oxide v0.7.4
  Downloaded serde v1.0.204
  Downloaded memchr v2.7.4
  Downloaded flate2 v1.0.31
  Downloaded log v0.4.22
  Downloaded serde_json v1.0.122
  Downloaded clap_builder v4.5.13
  Downloaded zerocopy v0.8.0-alpha.16
  Downloaded aho-corasick v1.1.3
  Downloaded chrono v0.4.38
  Downloaded serde_derive v1.0.204
  Downloaded syn v2.0.71
  Downloaded regex v1.10.5
  Downloaded nix v0.26.4
  Downloaded schemars v1.0.0-alpha.2
  Downloaded regex-syntax v0.8.4
  Downloaded proc-macro2 v1.0.86
  Downloaded num-traits v0.2.19
  Downloaded clap v4.5.13
  Downloaded serde_derive_internals v0.29.1
  Downloaded ryu v1.0.18
  Downloaded iana-time-zone v0.1.60
  Downloaded env_logger v0.11.5
  Downloaded clap_derive v4.5.13
  Downloaded bitflags v2.6.0
  Downloaded unicode-ident v1.0.12
  Downloaded unescaper v0.1.5
  Downloaded thiserror-impl v1.0.62
  Downloaded strsim v0.11.1
  Downloaded scopeguard v1.2.0
  Downloaded ref-cast v1.0.23
  Downloaded regex-automata v0.4.7
  Downloaded quote v1.0.36
  Downloaded itoa v1.0.11
  Downloaded humantime v2.1.0
  Downloaded dyn-clone v1.0.17
  Downloaded crc32fast v1.4.2
  Downloaded schemars_derive v1.0.0-alpha.2
  Downloaded libc v0.2.155
  Downloaded ref-cast-impl v1.0.23
  Downloaded pkg-config v0.3.30
  Downloaded libudev v0.3.0
  Downloaded heck v0.5.0
  Downloaded clap_lex v0.7.2
  Downloaded 60 crates (4.9 MB) in 0.99s
   Compiling proc-macro2 v1.0.86
   Compiling unicode-ident v1.0.12
   Compiling memchr v2.7.4
   Compiling libc v0.2.155
   Compiling utf8parse v0.2.2
   Compiling pkg-config v0.3.30
   Compiling autocfg v1.3.0
   Compiling serde v1.0.204
   Compiling libudev-sys v0.1.4
   Compiling quote v1.0.36
   Compiling num-traits v0.2.19
   Compiling aho-corasick v1.1.3
   Compiling syn v2.0.71
   Compiling anstyle-parse v0.2.5
   Compiling thiserror v1.0.62
   Compiling is_terminal_polyfill v1.70.1
   Compiling anstyle-query v1.1.1
   Compiling colorchoice v1.0.2
   Compiling cfg-if v1.0.0
   Compiling anstyle v1.0.8
   Compiling regex-syntax v0.8.4
   Compiling anstream v0.6.15
   Compiling regex-automata v0.4.7
   Compiling serde_json v1.0.122
   Compiling ref-cast v1.0.23
   Compiling serde_derive_internals v0.29.1
   Compiling regex v1.10.5
   Compiling bitflags v1.3.2
   Compiling log v0.4.22
   Compiling strsim v0.11.1
   Compiling iana-time-zone v0.1.60
   Compiling serde_derive v1.0.204
   Compiling thiserror-impl v1.0.62
   Compiling ref-cast-impl v1.0.23
   Compiling itoa v1.0.11
   Compiling ryu v1.0.18
   Compiling adler v1.0.2
   Compiling heck v0.5.0
   Compiling zerocopy v0.8.0-alpha.16
   Compiling clap_lex v0.7.2
   Compiling schemars v1.0.0-alpha.2
   Compiling clap_builder v4.5.13
   Compiling unescaper v0.1.5
   Compiling clap_derive v4.5.13
   Compiling miniz_oxide v0.7.4
   Compiling zerocopy-derive v0.8.0-alpha.16
   Compiling schemars_derive v1.0.0-alpha.2
   Compiling env_filter v0.1.2
   Compiling nix v0.26.4
   Compiling libudev v0.3.0
   Compiling crc32fast v1.4.2
   Compiling bitflags v2.6.0
   Compiling chrono v0.4.38
   Compiling scopeguard v1.2.0
   Compiling dyn-clone v1.0.17
   Compiling humantime v2.1.0
   Compiling env_logger v0.11.5
   Compiling serialport v4.4.0
   Compiling flate2 v1.0.31
   Compiling clap v4.5.13
   Compiling taptap v0.1.0 (/home/scor/Projects/taptap)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 53.67s
scor@Cloud ~/Projects/taptap $ cargo run observe --tcp 10.0.12.4
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.80s
     Running `target/debug/taptap observe --tcp 10.0.12.4`
scor@Cloud ~/Projects/taptap $
```
I can see it's trying to get some data from the RS-485 converter:
```
17:52:33.530906 IP (tos 0x0, ttl 64, id 55982, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.9.10.50222 > 10.0.12.4.7160: Flags [S], cksum 0x293c (incorrect -> 0xb096), seq 1219331598, win 64240, options [mss 1460,sackOK,TS val 1638359777 ecr 0,nop,wscale 7], length 0
17:52:33.550772 IP (tos 0x0, ttl 62, id 88, offset 0, flags [none], proto TCP (6), length 44)
    10.0.12.4.7160 > 10.0.9.10.50222: Flags [S.], cksum 0xa083 (correct), seq 2988335563, ack 1219331599, win 3000, options [mss 1460], length 0
17:52:33.550831 IP (tos 0x0, ttl 64, id 55983, offset 0, flags [DF], proto TCP (6), length 40)
    10.0.9.10.50222 > 10.0.12.4.7160: Flags [.], cksum 0x2928 (incorrect -> 0xc907), ack 1, win 64240, length 0
17:52:33.550958 IP (tos 0x0, ttl 64, id 55984, offset 0, flags [DF], proto TCP (6), length 40)
    10.0.9.10.50222 > 10.0.12.4.7160: Flags [F.], cksum 0x2928 (incorrect -> 0xc906), seq 1, ack 1, win 64240, length 0
17:52:33.552900 IP (tos 0x0, ttl 62, id 89, offset 0, flags [none], proto TCP (6), length 40)
    10.0.12.4.7160 > 10.0.9.10.50222: Flags [.], cksum 0xb83f (correct), ack 2, win 3000, length 0
17:52:33.553009 IP (tos 0x0, ttl 62, id 90, offset 0, flags [none], proto TCP (6), length 40)
    10.0.12.4.7160 > 10.0.9.10.50222: Flags [F.], cksum 0xb83e (correct), seq 1, ack 2, win 3000, length 0
17:52:33.553033 IP (tos 0x0, ttl 64, id 55985, offset 0, flags [DF], proto TCP (6), length 40)
    10.0.9.10.50222 > 10.0.12.4.7160: Flags [.], cksum 0x2928 (incorrect -> 0xc906), ack 2, win 64239, length 0
```
But that's it, there's nothing more I can see right now. The converter is set to 38400 8N1 no FlowCtrl. Can you please point me in the right direction? It has to be something simple, but I fail to recognize it...

**Comment by willglynn** [comment]
Hmm… could you attach a `.pcap` of that exchange? Do you see data when you `telnet 10.0.12.4 7160`?

**Comment by Scorcerer** [comment]
Hi!
Sure thing, here you go. When I connected to the gate via telnet I got a bunch of gibberish and then eventually my console locked up, but I think that's because of the amount of weird characters it was trying to display. I'll make a screenshot tomorrow when there will be some data on the wire.
Ah, I'm using this:
cargo 1.80.1 (376290515 2024-07-16)
to build the project.

**Comment by willglynn** [comment]
Got it, thanks! 

Well, `taptap` certainly appears to be closing the TCP client connection on purpose, which is… something. 🤔

**Comment by willglynn** [comment]
I wonder if the TCP serial device does not like half-closed clients: ab2455cf7e53f788289e97d2e756a30458dbf811. Please try `git fetch origin; git checkout do_not_half_close` and re-run.

**Comment by Scorcerer** [comment]
Awesome, it's not exiting now. Can't say more as it's night (still :P) but I'll get back to you tomorrow :)

**Comment by Scorcerer** [comment]
I left it running in a "screen" and it didn't log anything, but after restarting the process just now the data is flowing in :)
Thank you for your very quick response!

**Comment by willglynn** [comment]
Great! I'll merge this.

**Comment by willglynn** [comment]
I just released v0.1.1 containing this change.
